packages/host/app/components/operator-mode/container.gts

```diff
@@ -586,9 +602,76 @@ export default class OperatorModeContainer extends Component<Signature> {
   };
 
   @action updateSubmode(submode: Submode) {
+    switch (submode) {
+      case Submode.Interact:
+        this.updateCodePath(null);
+        break;
+      case Submode.Code:
+        let codePath = this.lastCardInRightMostStack
+          ? new URL(this.lastCardInRightMostStack.id + '.json')
+          : new URL(this.cardService.defaultURL + 'index.json');
+        this.updateCodePath(codePath);
+        break;
+      default:
+        throw assertNever(submode);
+    }
+
     this.operatorModeStateService.updateSubmode(submode);
   }
 
+  @action resetLoadFileError() {
+    this.loadFileError = null;
+  }
+
+  openFile = maybe(this, (context) => {
+    let realmURL = this.cardService.getRealmURLFor(this.codePath);
+    if (!realmURL) {
+      return undefined;
+    }
+
+    const realmPaths = new RealmPaths(realmURL);
+    const relativePath = realmPaths.local(this.codePath);
+    if (relativePath) {
+      return file(context, () => ({
+        relativePath,
+        realmURL: realmPaths.url,
+        onStateChange: (state) => {
+          if (state === 'not-found') {
+            this.loadFileError = 'File is not found';
+          }
+        },
+      }));
+    } else {
+      return undefined;
+    }
+  });
+
+  get codePath() {
+    return (
+      this.operatorModeStateService.state.codePath ??
+      (this.lastCardInRightMostStack
+        ? new URL(this.lastCardInRightMostStack.id + '.json')
+        : new URL(this.cardService.defaultURL + 'index.json'))
+    );
+  }
+
+  @action
+  updateCodePath(codePath: URL | null) {
+    this.operatorModeStateService.updateCodePath(codePath);
+    this.fetchCodeModeRealmInfo.perform();
+  }
```

Comment:
i'm confused by this--it looks like the container component has a codePath state, but also the operatorModeStateService has a codePath state. why do we have to track the code path as state in 2 different places? what happens when they are different?

Reply:
Updated here [Use codePath from operator mode service](https://github.com/cardstack/boxel/pull/571/commits/0f19726343a7b6185e7960aae0878210facb4b4e)